File: rust/src/group_anagrams.rs
//! This module was written to solve the "Group Anagrams" problem from LeetCode.
//!
//! Given an array of strings `strs`, group the anagrams together. You can return the answer in any order.
//! 
//! Examples:
//! Input: strs = ["eat","tea","tan","ate","nat","bat"]
//! Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
//! 
//! Input: strs = [""]
//! Output: [[""]]
//! 
//! Usage: group_anagrams(strs)

use std::collections::{BinaryHeap, HashMap};

/// Groups anagrams within a list of string values using HeapSorted char Vecs as keys for grouping the anagrams
pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
  let mut hash_groups: HashMap<Vec<char>, Vec<String>> = HashMap::new();

  for ele in strs {
    let ele_col: BinaryHeap<char> = ele.chars().collect();
    let sorted_ele = ele_col.into_sorted_vec();

      match hash_groups.get_mut(&sorted_ele) {
          Some(val) => {
            let ele_clone = ele.clone();
            val.push(ele_clone);
          },
          None => {
            let ele_clone = ele.clone();
            hash_groups.insert(sorted_ele, vec![ele_clone]);
          }
      }
  }

  hash_groups.into_values().collect()
}

File: py/group_anagrams.py
"""
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Examples:
Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

Input: strs = [""]
Output: [[""]]

Input: strs = ["a"]
Output: [["a"]]

Problem source: LeetCode

Usage: group_anagrams.py <comma_separated_strs>
"""

from typing import List
from collections import defaultdict
from utils import read_str_array
import sys

def group_anagrams(strs: List[str]) -> List[List[str]]:
  anagrams = defaultdict(list)

  for word in strs:
    anagrams["".join(sorted(word))].append(word)

  return list(anagrams.values())


if __name__ == "__main__":
  print(group_anagrams(read_str_array(sys.argv[1])))